### PR TITLE
fix: make set_more_retries() work when upstream_type is chash

### PR DESCRIPTION
**Pull request by zfs123**
chash will pick next if last doesn't work to enable set_more_retries().

### Pre-submission checklist:

* [x] Did you explain what problem does this PR solve? Or what new features have been added?
* [ ] Have you added corresponding test cases?
* [ ] Have you modified the corresponding document?
* [x] Is this PR backward compatible?

### CODE REVIEW

File: apisix/balancer/chash.lua [hunk 1]
```diff
@@ -55,6 +55,7 @@ end
 
 function _M.new(up_nodes, upstream)
     local str_null = str_char(0)
+    local last_server_index
```
Comment (spacewander) [inline apisix/balancer/chash.lua:58]: The `last_server_index` should be in ctx scope, not the upstream one. We should not share the same index across different requests to an upstream.

**Review by spacewander — CHANGES_REQUESTED** [review]

File: apisix/balancer/chash.lua [hunk 2]
```diff
@@ -68,8 +68,13 @@ function _M.new(up_nodes, upstream)
     return {
         upstream = upstream,
         get = function (ctx)
-            local chash_key = fetch_chash_hash_key(ctx, upstream)
-            local id = picker:find(chash_key)
+            local id
+            if ctx.balancer_try_count > 1 and ctx.last_server_index then
+                id, ctx.last_server_index = picker:next(ctx.last_server_index)
```
Comment (spacewander) [inline apisix/balancer/chash.lua:73]: Better to use name like `ctx.chash_last_server_index`

**Review by spacewander — CHANGES_REQUESTED** [review]

**Comment by spacewander** [comment]
Please write new test to verify the fix and make sure the tests are passed.

**Comment by moonming** [comment]
ping @zfs123

**Review by spacewander — APPROVED** [review]

**Review by tokers — APPROVED** [review]

**Pull request merged**